modules/kinesis/src/main/scala/com/snowplowanalytics/snowplow/enrich/kinesis/Sink.scala

```diff
@@ -224,6 +233,69 @@ object Sink {
       prre
     }
 
+  /** Describes an record-level error from trying to write a batch to kinesis */
+  private case class TryBatchResultError(count: Int, message: String)
+
+  /** The result of trying to write a batch to kinesis */
+  private case class TryBatchResult(
+    toRetry: Vector[PutRecordsRequestEntry],
+    throttledCount: Int,
+    otherErrors: Map[String, TryBatchResultError]
+  )
+
+  private object TryBatchResult {
+
+    implicit private def tryBatchResultErrorSemigroup: Semigroup[TryBatchResultError] =
+      new Semigroup[TryBatchResultError] {
+        override def combine(x: TryBatchResultError, y: TryBatchResultError): TryBatchResultError =
+          // Take the first message as an example, drop other messages.
+          TryBatchResultError(x.count + y.count, x.message)
+      }
+
+    implicit private def tryBatchResultMonoid: Monoid[TryBatchResult] =
+      new Monoid[TryBatchResult] {
+        override val empty: TryBatchResult = TryBatchResult(Vector.empty, 0, Map.empty)
+        override def combine(x: TryBatchResult, y: TryBatchResult): TryBatchResult =
+          TryBatchResult(x.toRetry ++ y.toRetry, x.throttledCount + y.throttledCount, x.otherErrors |+| y.otherErrors)
+      }
+
+    def build(records: List[PutRecordsRequestEntry], prr: PutRecordsResult): TryBatchResult =
+      if (prr.getFailedRecordCount.toInt =!= 0)
+        records
+          .zip(prr.getRecords.asScala)
+          .map {
+            case (orig, recordResult) =>
+              Option(recordResult.getErrorCode) match {
+                case None =>
+                  Monoid[TryBatchResult].empty
+                case Some("ProvisionedThroughputExceededException") =>
+                  TryBatchResult(Vector(orig), 1, Map.empty)
+                case Some(errorCode) =>
+                  TryBatchResult(Vector(orig), 0, Map(errorCode -> TryBatchResultError(1, recordResult.getErrorMessage)))
```

Comment:
You're right, it will always be InternalFailure (unless AWS one day returns something surprising and different).

I don't like this `Map[String, TryBatchResultError]` any more.  I'm only using it for one thing: logging a debug message.  It's too much complexity just for debugging.

I'm going to try out something different: I will try removing the Map entirely from the case class, but instead implement the debug logging closer to where we receive the raw aws response.